include/aws/common/cbor.h

```diff
@@ -0,0 +1,450 @@
+ * as it's not well supported cross languages.
+ *
+ * To be more specific, it will be encoded into integer/negative/float
+ * (Order with priority) when the conversation will not cause precision loss.
+ *
+ * @param encoder
+ * @param value value to encode.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_double(struct aws_cbor_encoder *encoder, double value);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_BYTESTRING value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1, the length of "from" will be encoded first and then the value of "from" will
+ * be followed.
+ *
+ * @param encoder
+ * @param from value to encode.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_bytes(struct aws_cbor_encoder *encoder, const struct aws_byte_cursor from);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_STRING value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1, the length of "from" will be encoded first and then the value of "from" will
+ * be followed.
+ *
+ * @param encoder
+ * @param from value to encode.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_string(struct aws_cbor_encoder *encoder, const struct aws_byte_cursor from);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_ARRAY_START value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1
+ *  The "number_entries" is the cbor data items should be followed as the content of the array.
+ * Notes: it's user's responsibility to keep the integrity of the array to be encoded.
+ *
+ * @param encoder
+ * @param number_entries The number of data item in array.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_array_start(struct aws_cbor_encoder *encoder, size_t number_entries);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_MAP_START value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1
+ *  The "number_entries" is the number of pair of cbor data items as key and value should be followed as the content of
+ * the map.
+ *
+ * Notes: it's user's responsibility to keep the integrity of the map to be encoded.
+ *
+ * @param encoder
+ * @param number_entries The number of data item in map.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_map_start(struct aws_cbor_encoder *encoder, size_t number_entries);
+
+/**
+ * @brief Encode a AWS_CBOR_TYPE_TAG value to "smallest possible" in encoder's buffer.
+ *  Referring to RFC8949 section 4.2.1
+ * The following cbor data item will be the content of the tagged value.
+ * Notes: it's user's responsibility to keep the integrity of the tagged value to follow the RFC8949 section 3.4
+ *
+ * @param encoder
+ * @param tag_number The tag value to encode.
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_tag(struct aws_cbor_encoder *encoder, uint64_t tag_number);
+
+/**
+ * @brief Encode a simple value AWS_CBOR_TYPE_NULL
+ *
+ * @param encoder
+ */
+AWS_COMMON_API
+void aws_cbor_encoder_write_null(struct aws_cbor_encoder *encoder);
```

Comment:
trivial/debatable: worth naming like `write_simple_null()` to group all the simples? Or just have like `write_simple(struct aws_cbor_encoder *encoder, uint64_t simple_value)`? (I'm not sure if user-defined simple values are a thing...)

I guess that would allow users to pass illegal simple values, but this API also allows users to encode their arrays and maps wrong too, so we could just naively forge ahead and encode whatever they say and that's at least consistent API-wise